src/bpfprogram.cpp

```diff
@@ -1,569 +1,99 @@
-    if (insn->code != (BPF_LD | BPF_DW | BPF_IMM)) {
-      LOG(ERROR) << "Cannot relocate insn code " << insn->code << " ld "
-                 << (insn->code & BPF_LD) << " dw " << (insn->code & BPF_DW)
-                 << " imm " << (insn->code & BPF_IMM);
-      throw std::invalid_argument("Unsupported relocated instruction");
-    }
-
-    auto *sym = &reinterpret_cast<const Elf64_Sym *>(symtab.data())[relsym];
-    auto symtype = ELF64_ST_TYPE(sym->st_info);
-    if (symtype == STT_FUNC || symtype == STT_SECTION) {
-      // We will append .text to code_ and we assume that sym->st_shndx
-      // corresponds to .text, therefore symbol value is an offset from
-      // code_.size().
-      //
-      // Relocate via direct instruction manipulation instead of the
-      // relocation entry for readability purposes.
-      //
-      // This is a hack. We shouldn't do this. However, we don't actually have
-      // the ELF section table to determine if the relocation actually refers
-      // to .text
-      if (text_offset_ == 0)
-        text_offset_ = code_.size();
-
-      // An offset to .text must be added when relocating instructions from the
-      // main program that point into .text.
-      // No offset is needed when relocating instructions from .text that point
-      // to another part of .text.
-      const auto text_relative_offset = (relsecname == ".rel.text")
-                                            ? 0
-                                            : text_offset_;
+  std::string mod = probe.path;
+  std::string fun = probe.attach_point;
 
-      auto target_insn = (text_relative_offset + sym->st_value + insn->imm) /
-                         sizeof(struct bpf_insn);
-      insn->src_reg = BPF_PSEUDO_FUNC;
-      insn->imm = (target_insn - insn_offset - 1); // jump offset
-    } else if (symtype == STT_OBJECT) {
-      std::string map_name = bpftrace_map_name(
-          reinterpret_cast<const char *>(strtab.data() + sym->st_name));
-
-      const auto &map = bytecode.getMap(map_name);
-      insn->src_reg = map.bpf_name() == to_string(MapType::MappedPrintfData)
-                          ? BPF_PSEUDO_MAP_VALUE
-                          : BPF_PSEUDO_MAP_FD;
-      insn->imm = static_cast<int32_t>(map.fd);
+  std::string btf_fun = probe.type == ProbeType::iter ? "bpf_iter_" + fun : fun;
+  if (btf.get_btf_id(btf_fun, mod) < 0) {
+    std::string msg = "No BTF found for " + mod + ":" + fun;
+    if (probe.orig_name != probe.name) {
+      // one attachpoint in a multi-attachpoint (wildcard or list) probe
+      // failed, print a warning but continue
+      LOG(WARNING) << msg << ", skipping.";
     } else {
-      LOG(ERROR) << "Relocation in " << relsecname << " type " << reltype
-                 << " sym " << relsym << " type " << symtype;
-      throw std::invalid_argument(
-          "Unsupported symbol type referenced in relocation");
+      // explicit match failed, fail hard
+      throw std::runtime_error(msg);
```

Comment:
Throw a FatalUserException instead?

Reply:
Good point. Also I missed to respect the new `missing_probes` config.